rfc/user-secrets.md

```diff
@@ -0,0 +1,186 @@
+
+It may be possible to store the Spinnaker authorization data for secrets in secrets metadata APIs, though the implementation details on how each secrets manager exposes metadata varies quite a bit.
+Since these access controls are not particularly sensitive, they don't _need_ to be encrypted with the secret, though this is the most direct way to deliver said access controls such that Fiat can understand them.
+
+This `UserSecretManager` API would replace the tentative use of `SecretManager` in the Clouddriver Account Management API.
+This API extends from the existing `SecretEngine` API, so it will integrate with existing engines fairly easily.
+
+Some other approaches considered here explored the possibility for using more sophisticated identity and access management integration in Spinnaker using a framework such as [SPIFFE](https://spiffe.io) which could potentially be used to identify subjects inside Spinnaker and for generating appropriate certificates and tokens.
+This approach was considered as the ultimate goal of using secrets in Spinnaker in the first place, though such an integration project would be much more expansive than a user secrets API.
+While the SPIFFE approach is a better idea in theory, it would still require all the relevant integration points with Spinnaker to also support the same framework.
+Integrating with external secrets managers provides more flexibility here as the secrets manager can be used for generating and rotating secrets for far more systems than are currently supported by SPIFFE.
+A similar approach was considered specific to Kubernetes using JWTs that both Kubernetes and Fiat could understand and exchange, though this is similarly limited by the number of integration points that support OpenID-Connect or other JWT-based auth systems.
+
+## Known Unknowns
+
+Some aspects of this feature still require more research.
+
+* Configuration strategy for how long to cache fetched secrets.
+This needs to be designed so that secrets can be rotated while also avoiding unnecessary load on the relevant secrets manager APIs.
+* How to configure multiple authentication identities for use with fetching secrets.
+This currently relies on having a single Spinnaker identity to route secrets requests through, though in a more dynamic environment, there may be multiple workload identities in Spinnaker.
+* How this would work in other cloud providers.
+Kork only includes `SecretEngine` implementations for AWS and GCP; implementing equivalent engines for other cloud providers is out of scope for this initial RFC.
+
+## Security, Privacy, and Compliance
+
+This feature is a security feature and thus must be carefully reviewed.
+The proposed architecture involves various Spinnaker services being granted read access to secrets along with those services keeping cached secrets from multiple tenants.
```

Comment:
Not a blocker, but I could see this causing rate limiting issues

Reply:
Potentially, yeah. I already updated my own custom `SecretEngine` implementation to configure some auto-retry with exponential backoff logic on API calls when being rate-limited. I'm thinking it could be useful to configure Spring caching on some of these beans, though that will require evaluating the security of encrypted cache stores like Redis for this purpose.